[Wallet] Remove CWalletDB* parameter from CWallet::AddToWallet

pstratem opened the pull request:
This is the last method to pass a CWalletDB

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -687,8 +687,10 @@ void CWallet::MarkDirty()
-bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFromLoadWallet, CWalletDB* pwalletdb)
+bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFromLoadWallet)
 {
+    CWalletDB walletdb(strWalletFile);
```
just curious: what is the reason to not keep a instance variable (within CWallet) that holds walletdb object? What are the benefits of constantly opening/instantiating the CWalletDB?

laanwj replied:
The scope is used to manage the lifetime of the transaction.

pstratem replied:
@jonasschnelli I will eventually get to that point, but it's going to take a significant amount of work to get there.

jonasschnelli commented on src/wallet/walletdb.cpp:
```diff
@@ -400,7 +400,7 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
                 wss.fAnyUnordered = true;
 
-            pwallet->AddToWallet(wtx, true, NULL);
+            pwallet->AddToWallet(wtx, true);
```
If you have 10'000 wallet transaction, this would now result in 10'000 x `CWalletDB walletdb(strWalletFile);`.
What about the performance impacts?

sipa commented:
I think the persistent CWalletDb object here is just there to prevent a flush to disk in between every added transaction, for performance reasons. Based on earlier pull requests by @pstratem around this, I think he intends to re-introduce a single CWalletDb* inside CWallet rather than reopening one all the time, after cleaning things up. Perhaps that will need to be done in the same PR, though?

dcousens commented:
> Perhaps that will need to be done in the same PR, though?

I think provided it is done before a release,  there is no issue.

MarcoFalke commented:
I think there is no advantage merging this small refactoring early. Might as well just add the second commit to address @jonasschnelli s and @sipa s feedback.

jonasschnelli commented:
Regardless of the persistent CWalletDB * I think this PR has value (refactoring). But I guess we should **not** change the `ReadKeyValue`/`AddToWallet` in that way that it instantiate a new CWalletDB on each wallet transaction load from the wallet.dat (which this PR currently does).

laanwj commented:
A refactoring shouldn't result in behavioral changes. I'd rather stick with ugly code somewhat longer than introduce a temporary performance regression.

dcousens commented:
>  introduce a temporary (but serious) performance regression.

At least then you'll have users reminding you :laughing:

pstratem commented:
I've split up AddToWallet into AddToWallet and LoadToWallet around the fFromLoadWallet flag.

This addresses the performance @jonasschnelli noticed and improves the API.

Commit:
Split CWallet::AddToWallet into AddToWallet and LoadToWallet.

This removes the fFromLoadWallet flag in AddToWallet.  These were already
effectively two methods.

Commit:
Remove CWalletDB* parameter from CWallet::AddToWallet

pstratem commented:
I swapped the order of the commits such that there's no performance issue at any point.

Commit:
Remove unused pwalletdb from CWallet::AddToWallet

pstratem commented:
import hashlib;hashlib.sha256("there is still a performance issue").hexdigest()

there isn't actually

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -2435,9 +2435,6 @@ bool CWallet::CommitTransaction(CWalletTx& wtxNew, CReserveKey& reservekey)
                 NotifyTransactionChanged(this, coin.GetHash(), CT_UPDATED);
             }
-
-            if (fFileBacked)
```
IMO the non-file-backend state is only used in UnitTests, right? We should also consider dropping it completely (in a later PR).

pstratem replied:
I'll be able to easily remove it once more of the logic is encapsulated.

jonasschnelli commented on src/wallet/rpcdump.cpp:
```diff
@@ -302,8 +302,7 @@ UniValue importprunedfunds(const UniValue& params, bool fHelp)
     LOCK2(cs_main, pwalletMain->cs_wallet);
 
     if (pwalletMain->IsMine(tx)) {
-        CWalletDB walletdb(pwalletMain->strWalletFile, "r+", false);
```
The third argument `fFlushOnClose` (`false` here) will be `true` after this PR (default value is `true`).
= before this PR `importprunedfunds` will not flush on close. After this PR it will.
But I guess this is not a problem.

pstratem replied:
Indeed this does change the default. There doesn't seem to be a good reason not to flush here.

jonasschnelli commented:
utACK 44d5267bfcd8abae5f0ae2a0f4052fc5a88acd2e

MarcoFalke commented:
utACK 44d5267

instagibbs commented:
tACK

MarcoFalke commented:
Seems 6246add is just  a plain rebase of 44d5267

re-utACK  6246add

pstratem commented:
@MarcoFalke indeed it is

jonasschnelli commented:
Looks good now!
Extensive Code Review ACK (6246add89a46523526a57b66bb87daf9413c69d7), nit: maybe squash commit.

pstratem commented:
@jonasschnelli I'd rather not squash, each commit is logically independent.

laanwj commented on src/wallet/wallet.cpp:
```diff
@@ -904,7 +909,7 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransaction& tx, const CBlock* pbl
             // this is safe, as in case of a crash, we rescan the necessary blocks on startup through our SetBestChain-mechanism
             CWalletDB walletdb(strWalletFile, "r+", false);
```
This can go

pstratem replied:
Indeed this can go. I'll run a benchmark of master and this pr doing a rescan with a wallet that has an enormous number of transactions and keys.

sipa commented:
utACK c04225e7fdb9651ae2f97ed0e3f9601205ea858b

pstratem commented:
@laanwj I was having trouble getting consistent benchmark results from -rescan (+-30 seconds on runs taking 160-200 seconds).

So I just added an fFlushOnClose parameter so it's now clearly doing the same thing as before the change.

sipa commented:
reutACK 5723bb44ce2c6bb14114aa7f211160702a47ac91

MarcoFalke commented:
> indeed it is

I prefer no rebase when it is not necessary. Reviewers always have to reconstruct the history locally each time a rebase is done.

pstratem commented:
Indeed I generally agree, but I was already squashing adding the
fFlushOnClose flag.

Whenever possible I strongly prefer that each commit produces a complete
working build.

On Jul 31, 2016 3:56 AM, "MarcoFalke" <notifications@github.com> wrote:

> indeed it is
>
> I prefer no rebase when it is not necessary. Reviewers always have to
> reconstruct the history locally each time a rebase is done.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/8152#issuecomment-236423575>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAl4Q5FO3twJijzI_pOno3iNIKzaHvLfks5qbH8-gaJpZM4IvJOB>
> .
>

MarcoFalke commented:
With rebase I meant changing the commit which the pull is based on,
not squashing.


reACK 5723bb44ce2c6bb14114aa7f211160702a47ac91
